zcutil/fetch-params.sh

```diff
@@ -8,25 +8,47 @@ SPROUT_PKEY_NAME='sprout-proving.key'
 SPROUT_VKEY_NAME='sprout-verifying.key'
 SPROUT_PKEY_URL="https://z.cash/downloads/$SPROUT_PKEY_NAME"
 SPROUT_VKEY_URL="https://z.cash/downloads/$SPROUT_VKEY_NAME"
+SPROUT_PKEY_IPFS="/ipfs/QmZKKx7Xup7LiAtFRhYsE1M7waXcv9ir9eCECyXAFGxhEo/$SPROUT_PKEY_NAME"
+SPROUT_VKEY_IPFS="/ipfs/QmZKKx7Xup7LiAtFRhYsE1M7waXcv9ir9eCECyXAFGxhEo/$SPROUT_VKEY_NAME"
 
 SHA256CMD="$(command -v sha256sum || echo shasum)"
 SHA256ARGS="$(command -v sha256sum >/dev/null || echo '-a 256')"
 
+# disable ipfs support with ZC_DISABLE_IPFS=1
+ZC_DISABLE_IPFS="${ZC_DISABLE_IPFS:-}"
+IPFSCMD="$(command -v ipfs)"
+
 function fetch_params {
     local url="$1"
-    local output="$2"
+    local ipfspath="$2"
+    local output="$3"
     local dlname="${output}.dl"
-    local expectedhash="$3"
+    local expectedhash="$4"
 
     if ! [ -f "$output" ]
     then
-        echo "Retrieving: $url"
-        wget \
-            --progress=dot:giga \
-            --output-document="$dlname" \
-            --continue \
-            --retry-connrefused --waitretry=3 --timeout=30 \
-            "$url"
+        if ! [ -z "$IPFSCMD" ] && [ -z "$ZC_DISABLE_IPFS" ]; then
+            echo "Retrieving: $ipfspath"
+
+            if ! ipfs get --output "$dlname" "$ipfspath"; then
+                echo "Fallback: $url"
+                wget \
+                    --progress=dot:giga \
+                    --output-document="$dlname" \
+                    --continue \
+                    --retry-connrefused --waitretry=3 --timeout=30 \
+                    "$url"
```

Comment:
This duplicates the wget command and parameters.

I'd prefer that each retrieval mechanism be factored out into its own function; then the configuration and fallback logic will be more readable, and other mechanisms could be added in future more easily.

Reply:
Agreed. I've added a new patch that introduces `fetch_wget` and `fetch_ipfs`. I'm not really happy with the way the fallback works, but I'm not sure how to implement this cleanly in bash. Input welcome!

I'm basically looking for something similar to this:
```
if should try ipfs
    if ipfs worked
        goto done

if should try some_future_fetch
    if some_future_fetch worked
        goto done

wget

done:
    verify files